### PR TITLE
intel_adsp: device power management

### PR DESCRIPTION
**Pull request by tmleman**
Setting the DMIC and SSP devices into OFF state after initialization as this more closely matches the actual state of the device.

### CODE REVIEW

File: drivers/dai/intel/dmic/dmic.c [hunk 1]
```diff
@@ -876,14 +887,7 @@ const struct dai_driver_api dai_dmic_ops = {
...
 static int dai_dmic_initialize_device(const struct device *dev)
 {
-	IRQ_CONNECT(
-		DT_INST_IRQN(0),
-		IRQ_DEFAULT_PRIORITY,
-		dai_dmic_irq_handler,
-		DEVICE_DT_INST_GET(0),
-		0);
-
-	pm_device_init_suspended(dev);
```
Comment (softwarecki) [inline drivers/dai/intel/dmic/dmic.c:886]: Co call to the new `dai_dmic_irq_init` function?

Comment (tmleman) [reply drivers/dai/intel/dmic/dmic.c:886]: Function is called inside of the dmic_pm_action callback. During the transition from OFF state to the SUSPEND (when power domain is powered-up).

**Review by softwarecki — DISMISSED** [review]

**Review by kfrydryx — DISMISSED** [review]

**Comment by nashif** [comment]
@celoin please review

**Comment by tmleman** [comment]
@juimonen are there power domains on CAVS?

File: drivers/dai/intel/dmic/dmic.c [hunk 2]
```diff
@@ -841,6 +841,16 @@ static int dai_dmic_remove_wrapper(const struct device *dev)
...
 
+static inline void dai_dmic_irq_init(void)
+{
+	IRQ_CONNECT(
+		DT_INST_IRQN(0),
+		IRQ_DEFAULT_PRIORITY,
+		dai_dmic_irq_handler,
+		DEVICE_DT_INST_GET(0),
+		0);
+}
```
Comment (ceolin) [inline drivers/dai/intel/dmic/dmic.c:852]: This is not making too much sense to me. That is a static configuration, why don't you do it in `dai_dmic_initialize_device` ?

File: drivers/dai/intel/dmic/dmic.c [hunk 3]
```diff
@@ -850,9 +860,10 @@ static int dmic_pm_action(const struct device *dev, enum pm_device_action action
 	case PM_DEVICE_ACTION_RESUME:
 		dai_dmic_probe_wrapper(dev);
 		break;
-	case PM_DEVICE_ACTION_TURN_OFF:
 	case PM_DEVICE_ACTION_TURN_ON:
-		/* All device pm is handled during resume and suspend */
+		dai_dmic_irq_init();
+		break;
```
Comment (ceolin) [inline drivers/dai/intel/dmic/dmic.c:865]: You are not doing anything special here, just do the same thing you do for `PM_DEVICE_ACTION_TURN_OFF` and move that device initialization to the init funciton. Note that this state is only triggered when the power domain this device belongs  is resumed.

**Review by ceolin — CHANGES_REQUESTED** [review]
There are static initialization that does not seem to be part of the device power management.

**Review by ceolin — APPROVED** [review]

**Review by abonislawski — APPROVED** [review]

**Review by marcinszkudlinski — APPROVED** [review]

**Review by softwarecki — APPROVED** [review]

**Pull request merged**